Observer/CheckoutSubmitBefore.php

```diff
@@ -0,0 +1,207 @@
+     * @param  bool   $blIsAddresscheck
+     * @return string
+     */
+    protected function getConfigParam($sParam, $blIsAddresscheck = false)
+    {
+        $sGroup = 'creditrating';
+        if ($blIsAddresscheck === true) {
+            $sGroup = 'address_check';
+        }
+        return $this->consumerscoreHelper->getConfigParam($sParam, $sGroup, 'payone_protect');
+    }
+
+    /**
+     * Determine if creditrating is needed
+     *
+     * @param  Quote $oQuote
+     * @return bool
+     */
+    protected function isCreditratingNeeded(Quote $oQuote)
+    {
+        if ((bool)$this->getConfigParam('enabled') === false) {
+            return false;
+        }
+
+        if ($this->getConfigParam('integration_event') != 'after_payment') {
+            return false;
+        }
+
+        $oMethodInstance = $oQuote->getPayment()->getMethodInstance();
+
+        $sPaymentCode = $oMethodInstance->getCode();
+        $sPaymentTypesToCheck = $this->getConfigParam('enabled_for_payment_methods');
+        $aPaymentTypesToCheck = explode(',', $sPaymentTypesToCheck);
+        if (array_search($sPaymentCode, $aPaymentTypesToCheck) === false) {
+            return false;
+        }
+
+        if ($oMethodInstance->getInfoInstance()->getAdditionalInformation('payone_boni_agreement') === false) {
+            return false;// agreement checkbox was not checked by the customer
+        }
+
+        $dTotal = $oQuote->getGrandTotal();
+        $dMin = $this->getConfigParam('min_order_total');
+        $dMax = $this->getConfigParam('max_order_total');
+        if (is_numeric($dMin) && is_numeric($dMax) && ($dTotal < $dMin || $dTotal > $dMax)) {
+            return false;
+        }
+
+        if ((bool)$this->getConfigParam('sample_mode_enabled')
+                && !empty($this->getConfigParam('sample_mode_frequency'))
+                && $this->consumerscoreHelper->isSampleNeeded() === false) {
+            return false;
+        }
+
+        return true;
+    }
+
+    protected function isPaymentApplicableForScore(Quote $oQuote, $sScore)
+    {
+        if ($sScore == 'G') {
+            return true;
+        }
+
+        $sPaymentCode = $oQuote->getPayment()->getMethodInstance()->getCode();
+
+        $aYellowMethods = $this->consumerscoreHelper->getAllowedMethodsForScore('Y');
+        $aRedMethods = $this->consumerscoreHelper->getAllowedMethodsForScore('R');
+
+        if ($sScore == 'Y' && (array_search($sPaymentCode, $aYellowMethods) !== false ||
+                array_search($sPaymentCode, $aRedMethods) !== false)) {
+            return true;
+        } elseif ($sScore == 'R' && array_search($sPaymentCode, $aRedMethods) !== false) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Filter payment methods by the creditrating result if applicable
+     *
+     * @param  AddressInterface $oBilling
+     * @return void
+     */
+    protected function getScoreByCreditrating(AddressInterface $oBilling)
```

Comment:
This method is too complex, please refactor: https://scrutinizer-ci.com/g/PAYONE-GmbH/magento-2/inspections/f57e816f-b8e5-4e29-b0a2-aa3d5399368d/code-structure/operation/Payone%5CCore%5CObserver%5CCheckoutSubmitBefore%3A%3AgetScoreByCreditrating